### ISSUE TITLE
Make some 7.3 Experimental Feature Stable

### ISSUE THREAD

**Issue by SteveL-MSFT** [PowerShell/PowerShell#17962]
### Summary of the new feature / enhancement

The current ExperimentalFeatures in 7.3:

```output
Name                                Enabled Source                              Description
----                                ------- ------                              -----------
PSAMSIMethodInvocationLogging          True PSEngine                            Provides AMSI notification of .NET method invocations.
PSAnsiRenderingFileInfo                True PSEngine                            Enable coloring for FileInfo objects
PSCleanBlock                           True PSEngine                            Add support of a 'Clean' block to functions and script cmdlets for easy resou…
PSCommandNotFoundSuggestion            True PSEngine                            Recommend potential commands based on fuzzy search on a CommandNotFoundExcept…
PSExec                                 True PSEngine                            Add 'exec' built-in command on Linux and macOS
PSLoadAssemblyFromNativeCode           True PSEngine                            Expose an API to allow assembly loading from native code
PSNativeCommandArgumentPassing         True PSEngine                            Use ArgumentList when invoking a native command
PSNativeCommandErrorActionPreferen…    True PSEngine                            Native commands with non-zero exit codes issue errors according to $ErrorActi…
PSRemotingSSHTransportErrorHandling    True PSEngine                            Removes the SSH remoting transport stdErr stream message handling as terminat…
PSStrictModeAssignment                 True PSEngine                            Add support of setting Strict-Mode with Invoke-Command
PSSubsystemPluginModel                 True PSEngine                            A plugin model for registering and un-registering PowerShell subsystems
```

### Proposed technical implementation details (optional)

Current discussion within the PS-Committee is not complete, but so far we've decided:

Make stable:

`PSAMSIMethodInvocationLogging` - confirm with AMSI team that the API is done and public
`PSAnsiRenderingFileInfo` - make stable to get more user feedback.  Since the decoration is user configurable, users can turn it off completely or adjust.  The main criteria here is to enable this capability and less about the default colors which are user configurable.  We appreciate the feedback (although would have liked it to come earlier) and agreed that since formatting changes are not breaking changes, we should proceed with the current defaults and we can continue discussing with the community potential changes for the defaults in 7.4 previews.
`PSCleanBlock` - an opt-in feature that's been available for awhile and commit to fixing bugs should they come up
`PSExec` - provide consistency with expectations on non-Windows platforms
`PSNativeCommandArgumentPassing` - we've received good feedback on this which greatly improves the experience on non-Windows and Windows specific issues have been addressed, we agreed to take this for 7.3 and continue to address issues into 7.4 as needed
`PSRemotingSSHTransportErrorHandling` - this addresses a major issue with SSH remoting where native command use stderr as information and not an error which previously would kill the connection, so we want to take this fix

Keep Experimental:

`PSCommandNotFoundSuggestion` - pending on rewrite of Suggestion framework to make it work correctly
`PSLoadAssemblyFromNativeCode` - keep as experimental to see if there are more users that want/need this type of feature
`PSNativeCommandErrorActionPreference` - keep as experimental as we don't have any data if users have used this feature given that it's opt-in by default, consider making it opt-out in 7.4
`PSSubsystemPluginModel` - re-discussed within the Committee and agreed to keep this experimental, but support Az.Predictor dependency

Remove as feature:

`PSStrictModeAssignment` - Remove as feature since it's not enabled for remoting and is not very useful when limited to local usage

We will review the other ones at next meeting.

**Comment by kasini3000** [comment]
I think there will be over 10,000 people looking at this issue. Users may not know powershell well, but are interested in new features. 
Could you be more specific about the description? Like #15862, thanks

**Comment by SteveL-MSFT** [comment]
You mean descriptions of the experimental features?  We have official docs here: https://docs.microsoft.com/en-us/powershell/scripting/whats-new/what-s-new-in-powershell-73?view=powershell-7.2

**Comment by iSazonov** [comment]
LGTM without PSSubsystemPluginModel.

**Comment by jhoneill** [comment]
> `PSAnsiRenderingFileInfo` - make stable to get more user feedback. Since the decoration is user configurable, users can turn it off completely or adjust

Given all the things I'm record saying about the application of colours, you won't be surprised that I want to advise caution on this point. 
In particular: the choice of colours and file extensions they are applied to up to this point seems arbitrary (IIRC ZIPs were formatted as error / danger). Possibly there has been consultation that I'm not aware of, but it feels like choices based on personal tastes rather than usability research.   
If the _ability to format_ were turned on but the default was not to apply any, users who want specific colours can still apply them. My guess is that making it opt-in would see quite a low take-up, opt-out will see it used by people who find the colours unhelpful, but neither find out how to turn them off, nor complain.

**Comment by SteveL-MSFT** [comment]
@jhoneill the colors chosen (with exception of directories) were from existing colors already used which don't have a contrast concern for accessibility.  I've been asking for suggestions on the defaults, but have not received any.  So if you have any specific suggestions we can discuss those.  Having it enabled, but without decoration only makes it harder to discover.

**Comment by dkaszews** [comment]
Ad colors: I'm quite satisfied with the defaults, as they closely match bash. And while we should absolutely not treat "we should be like bash" as the ultimate goal, opening two windows side by side and seeing both directory listings look similar was satisfying. Plus, it allows us to leverage existing user expectations from a well established color design.

**Comment by MartinGC94** [comment]
I've been using my own file formatting file to get lengths in human readable format for ages so I haven't noticed the change.  
Looking at them now I think the file and SymbolicLink colors are good but I strongly dislike the directory color. It sticks out way too much and considering how the Windows console doesn't handle background colors + resizing very well I think it's a bad idea to make this the default experience. Just look at this:
![image](https://user-images.githubusercontent.com/42123497/186723426-e5889bb8-5aa0-43a1-8945-ea67ab6b9ca4.png)

The BrightBlue color is much nicer IMO.

**Comment by dkaszews** [comment]
Yeah, agreed, sticking to foreground colors is a better default, that's why I set it myself to foreground blue. However what @MartinGC94 shows looks like a bug I cannot replicate, but that could be fixed by always making sure to reset color after filename

![Screenshot_20220825-190242~2](https://user-images.githubusercontent.com/26599495/186726662-b995b8eb-0b97-456c-97cd-08f1d71292c7.jpg)

**Comment by SteveL-MSFT** [comment]
@dkaszews the default colors, with the exception of directories, was borrowed from existing Linux colors.  The default color for directories for Linux would not pass the accessibility requirement for contrast which is why I chose a white foreground on blue background.

I think the question is not so much a preference (as the colors/decoration can all be user configured), but if the defaults are acceptable by the majority of users.  Some people may not want any color, but instead use italics, bold, or underline but those may not be good defaults.

**Comment by dkaszews** [comment]
@SteveL-MSFT I trust you have consulted accessibility team to pick the best option, and the fact this is easily configurable by users pretty much cuts off any discussion.

Small question since I have your attention and I recall you were the author of the MR. Any plans to make type coloring fully generic? This introduced ability to set color of `FileInfo` based on its properties, I can imagine for example ability to color `MatchInfo`s consequitive groups (and not just the entire match) useful, let's say first is blue, second is green, etc. I know there is some way to do it via xml, but this API looks nicer.

**Comment by SteveL-MSFT** [comment]
@dkaszews color options for `MatchInfo` would have to be an entirely new feature.  Feel free to open a new issue to discuss the idea and would be helpful to have proposal on how it would be exposed to users as well as use cases showing its utility.

**Comment by jhoneill** [comment]
@SteveL-MSFT  well, "already used" might mean the "explosion in a paint factory"  that is `ls` on linux - a paradigm case of runaway UI changes by people with great expertise in C, but little in usability, resulting in bad experience.  
My ubuntu/wsl OOB config applies 18 different combinations of colours to 130 file types. What proportion of Linux users know all 18?  As someone who doesn't live in bash, I have no idea why some directories are blue text on black, and others are have a green background, nor what yellow background, or cyan-text on a red background are meant to convey.  Extra information is put front of me but I can't decode it.

I don't think Windows explorer is unusual keeping all files in one colour, I _think_ most GUI file managers do the same (ready to be corrected). In VS Code colour is used for source control status, not type. And AIUI `ls` uses colour because it defaults to showing many names per line:  _without_ `-l`  colour is the only clue what you can or can't run, or cd into.     
But DIR going all the way back to the 1980s has marked directories (even in wide mode) and file extensions told you what you could run, so had no need of colour.  

> we should absolutely not treat "we should be like bash" as the ultimate goal

Absolutely. I also think we should feel able to say that bash introduced something for a reason that doesn't apply here, and it grew into a monster (18 colour combos and 130 file types) that we want to avoid.  Shells _should_ learn from each other, and that includes the mistakes of others.

Executables and directories are being coloured not because it is needed, but because that's what people coming to pwsh with a  linux background have grown used to. I can see them complaining that displaying directories in inverse is "stupid" and wanting to know "why can't they be in blue" (which I see running  `ls  /usr` in bash on wsl - accessibility not being something linux spend much time on :-) )  But special colours for files with a PowerShell extension like ps1xml or psd1? That's just colour for colour's sake.  

I've tried to explain my accessibility problem and I don't know if the picture below helps to show what it is going on in my brain when this is in front of me.
<img width="615" alt="Untitled" src="https://user-images.githubusercontent.com/16209593/186730740-aa25192e-6eaa-459d-ac2d-ebf932a45d38.png">

I can't _just read_ the stuff I want to read, with the visual hubbub of all those colours it takes extra concentration and in some cases I need to put my hand over the screen to stop my eye being dragged where I don't want it to be.  Possibly that disqualifies me from commenting, as the answer I seem to get is "we like colours, you can turn yours off".

**Comment by dkaszews** [comment]
@jhoneill My argument for coloring a little more than users might like is to show that it is possible. Personally I didn't find coloring by extension useful and especially did not like the default colors, but it was nice to learn such a thing is possible. All I had to do is add `$PSStyle.FileInfo.Extensions.Clear()` to my profile to disable it, so I believe as long as we provide customizability, the discoverability of the feature is more important than defaults.

**Comment by jhoneill** [comment]
> @jhoneill My argument for coloring a little more than users might like is to show that it is possible. Personally I didn't find coloring by extension useful and especially did not like the default colors, but it was nice to learn such a thing is possible. All I had to do is add `$PSStyle.FileInfo.Extensions.Clear()` to my profile to disable it, so I believe as long as we provide customizability, the discoverability of the feature is more important than defaults.

That's a valid view. It works when people are curious. You or I look at it and say "Neat you _can_ apply formatting, and the formatting I want is... _none_, I'll set that." I don't have the same curiosity when I use bash on wsl. I only found how it set the 18 colour combos for 130 file types in replying above. I _could_ make a bash look a bit more what I expect  but I _don't_: I sit their scowling at it for the the defaults that I can't be bothered to change (even the three colour bash prompt). I don't want people who are dragging their feet about moving off Windows PowerShell feeling about pwsh 7.3 the way I feel about bash. 

Set against that, I don't want people who currently use bash feeling that getting the look they want is a battle. Pwsh doing different things on Windows and Linux is one answer to that, but I think that has as many negatives as positives.  
And I do understand the desire _not_ to keep formatting secret after introducing it. But look at what I said above `ls` defaults to wide view, only shows names, and needs colours so you what is a directory. In CMD dir /w puts [] around a directory so you can see pick directories out, so the new colours might not be needed in normal view but they going to help wide view right ? 

<img width="716" alt="image" src="https://user-images.githubusercontent.com/16209593/187030003-c981e70c-2713-467b-a335-a62083f7abeb.png">

No, actually **the place where colour _would_ benefit pwsh, doesn't use it** (list format doesn't use it either)

@SteveL-MSFT  I did experiment with better defaults
![image](https://user-images.githubusercontent.com/16209593/187029178-7812462e-4e0a-4416-8475-913b5692cc0e.png)

1. This is Windows terminal with its interpretation  of the different colours and PWSH 7 on Windows has a dark blue background for me PowerShell 5 has traditional blue, Pwsh 6,  cmd and anything linux is black 
2. I think the group header block should be same format as the TableHeader 
3. I have tried bright white and bold but it is only 242/255  compared with the bright gray for normal being 204/255 which isn't really enough contrast. 
4. I tried directories in Normal colour and italic. It gets a "maybe" from me it removes the look-at-me look-at-me look-at-me  of having them in inverse, and it does say something is going on.
5. PS1 files formatted as executables makes more more sense to me than all kinds of PowerShell files having a common format. 
6. Executables in green but not **BOLD GREEN** turns the attention grabbing down a notch. 


There are 4 things which make me ask if you are **sure** this is ready.

1. The default colours feeling a bit random. Once they become the norm, changing them will be harder. 
2. Not setting the colour of the grouping custom control to match the table headers
3. Using new properties, LengthString and LastWriteTimeString,  but not providing a way to format them (e.g. I would like thousand separators , or might like KB, MB, GB, TB etc), unix-die-hards might want dates to change format depending age as they are used to, 
4. Only using the new NameString, LengthString and LastWriteString in table formats. In a perfect world I want formatting OFF for table view and ON for wide view.

**Comment by SteveL-MSFT** [comment]
@jhoneill I see your complaints, but don't understand exactly what you're asking for.  If your main complaint is the red for archives, then one option is that we simply remove archives from the default list.  `.ps1` is executable (within PowerShell) so should that be green instead?  The intent of the defaults was a discovery mechanism to show users how they can customize it themselves.  Having the hashtable for extensions empty, by default, is an option, but it won't be obvious to users how to extend or customize it.

**Comment by jhoneill** [comment]
@SteveL-MSFT   I don't really think saying apply what you do to names in Table views should also apply to wide views **IS** additional. I think when the ps1xml files were updated - it looks like some is in  `src\System.Management.Automation\engine\TypeTable_Types_Ps1Xml.cs` and some is in `src\System.Management.Automation\FormatAndOutput\DefaultFormatters\FileSystem_format_ps1xml`  - whoever made the changes was only thinking about the default / table view and didn't finish the job.

The  `LengthString` and `LastWriteTimeString`  hidden code properties have been added and they are used in the formatting. If you're saying that having them dormant and adding stuff to PSStyle to light them up is something that needs to come later in order to ship something worthwhile soon, it's a compromise I can live with, provide the intent to do it is there.   

Header groups, don't belong to the file experimental feature. It's not clear to me if that should a distinct another setting in PSStyle (my hunch is it should even if the default value matches column headings) and then I think whatever applies formatting to the column headings should apply it.  Probably the smart thing to do on this is to open an issue or a discussion for this one. 

I'm trying to help with the selection of good defaults, but I'm very aware that the way my brain reacts to a printed page or a screen is unusual. I don't want rail too much about how bad directories in inverse text is **for me** if it is helpful to almost everyone else, especially as I can change it. I think we've got to agreement on pwsh files /executables. (I'd use all the extensions in %pathext% + .ps1). 

Whether formatting compressed files adds anything is debatable. One of the things a preview is for is to get feedback on proposed changes to the default experience, so I think at some point you do have to say, "yes we are sure now.". That will get more data on how it's received, and how well it works for people.
